### PR TITLE
Add s3 storage path interfaces

### PR DESCRIPTION
**Pull request by s-vitaliy**
Part of #61 

## Scope

Implemented:
- Added basic interfaces related to the storage interaction
- Added unit tests

## Checklist

- [ ] GitHub issue exists for this change.
- [ ] Unit tests added and they pass.
- [ ] Line Coverage is at least 80%.
- [ ] Review requested on `latest` commit.

### CODE REVIEW

**Comment by s-vitaliy** [comment]
Not sure if we need the StorageBlobWriter trait here, but let it be.

File: framework/arcane-framework/.idea/modules/arcane-framework.arcane-framework-build.iml [hunk 1]
```diff
@@ -195,4 +195,9 @@
       </library>
     </orderEntry>
   </component>
+  <component name="SbtModule">
```
Comment (george-zubrienko) [inline framework/arcane-framework/.idea/modules/arcane-framework.arcane-framework-build.iml:198]: I believe this is missing from .gitignore. Please fix

File: framework/arcane-framework/build.sbt [hunk 2]
```diff
@@ -5,5 +5,20 @@ ThisBuild / scalaVersion := "3.6.1"
 lazy val root = (project in file("."))
   .settings(
     name := "arcane-framework",
-    idePackagePrefix := Some("com.sneaksanddata.arcane.framework")
+    idePackagePrefix := Some("com.sneaksanddata.arcane.framework"),
```
Comment (george-zubrienko) [inline framework/arcane-framework/build.sbt:8]: you are missing `scalaVersion` - should be 3.5
(let's not use Scala 2 - v3 has been out for a while, so I suggest we go with it right away)
https://www.scala-sbt.org/1.x/docs/Howto-Scala.html
https://www.scala-lang.org/download/all.html

File: framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
+package com.sneaksanddata.arcane.framework
+package services.storage.base
+
+import services.storage.exceptions.BlobStorageException
```
Comment (george-zubrienko) [inline framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala:4]: these are all models. So mb simply `main/scala/models/...`?

File: framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import zio.ZIO
+
+import java.net.URL
+
+/**
+ * A trait that defines the interface for writing to a blob storage.
+ *
+ * @tparam PathType The type of the path to the blob.
+ */
+trait BlobStorageWriter[PathType <: BlobPath] {
```
Comment (george-zubrienko) [inline framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala:17]: update syntax to Scala 3

File: framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  def removeBlob(blobPath: PathType, data: String): ZIO[Any, BlobStorageException, Unit]
+
+  /**
+   * Gets the URI of the blob at the given path.
+   *
+   * @param blobPath The path to the blob.
+   * @param data The data to get.
+   * @return The URI of the blob.
+   */
+  def getBlobUri(blobPath: PathType, data: String): ZIO[Any, BlobStorageException, URL]
```
Comment (george-zubrienko) [inline framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala:51]: I wonder why are these ZIO types and not just Futures?

File: framework/arcane-framework/src/main/scala/services/storage/base/StorageBlobReader.scala [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import services.storage.models.base.BlobPath
+
+import zio.ZIO
+
+/**
+ * A trait that defines the interface for reading from a blob storage.
+ *
+ * @tparam PathType The type of the path to the blob.
+ */
+trait StorageBlobReader[PathType <: BlobPath] {
```
Comment (george-zubrienko) [inline framework/arcane-framework/src/main/scala/services/storage/base/StorageBlobReader.scala:14]:
```suggestion
trait BlobStorageReader[PathType <: BlobPath] {
```

File: framework/arcane-framework/src/main/scala/services/storage/models/amazon/AmazonS3StoragePath.scala [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    */
+  def isAmazonS3Path(hdfsPath: String): Boolean = hdfsPath.matches(matchRegex)
+
+  /**
+    * Creates an [[AmazonS3StoragePath]] from the given HDFS path.
+    *
+    * @param hdfsPath The HDFS path.
+    * @return The [[AmazonS3StoragePath]].
+    */
+  def fromHdfsPath(hdfsPath: String): Try[AmazonS3StoragePath] = {
```
Comment (george-zubrienko) [inline framework/arcane-framework/src/main/scala/services/storage/models/amazon/AmazonS3StoragePath.scala:54]: you need to define `apply` for the method to be a valid monad. I think this method looks like a good candidate? 
it acts as a default constructor:
val myPath = AmazonS3StoragePath("s3a://...")
You can also provide overloads for `apply` AND apply quality checks in it - like the one you do in `isAmazonS3Path`

File: framework/arcane-framework/src/test/scala/services/storage/models/amazon/AmazonS3StoragePath.scala [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.prop.TableDrivenPropertyChecks.forAll
+import org.scalatest.prop.Tables.Table
+
+class AmazonS3StoragePath extends AnyFlatSpec {
+
+  "AmazonS3StoragePath" should "be able to parse correct path" in {
+    val path = "s3a://bucket/key"
+    val parsed = AmazonS3StoragePath.fromHdfsPath(path)
+    assert(parsed.isSuccess)
```
Comment (george-zubrienko) [inline framework/arcane-framework/src/test/scala/services/storage/models/amazon/AmazonS3StoragePath.scala:11]: these are .net/java style asserts. I'd suggest we use [matchers](https://www.scalatest.org/user_guide/using_matchers) that look like `result should be 1`
Please review https://stackoverflow.com/questions/61528075/scalatest-combine-assertions
and this one ofc https://www.scalatest.org/user_guide/using_matchers

**Review by george-zubrienko — COMMENTED** [review]
Work in progress - let's sync on Monday :)

File: framework/arcane-framework/src/test/scala/services/storage/models/amazon/AmazonS3StoragePath.scala (hunk 8 above)
Comment (s-vitaliy) [reply framework/arcane-framework/src/test/scala/services/storage/models/amazon/AmazonS3StoragePath.scala]: Fixed

File: framework/arcane-framework/src/main/scala/services/storage/models/amazon/AmazonS3StoragePath.scala (hunk 7 above)
Comment (s-vitaliy) [reply framework/arcane-framework/src/main/scala/services/storage/models/amazon/AmazonS3StoragePath.scala]: Fixed

File: framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala (hunk 5 above)
Comment (s-vitaliy) [reply framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala]: Fixed

File: framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala (hunk 4 above)
Comment (s-vitaliy) [reply framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala]: done

File: framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala (hunk 3 above)
Comment (s-vitaliy) [reply framework/arcane-framework/src/main/scala/services/storage/base/BlobStorageWriter.scala]: Removed the exception at all

File: framework/arcane-framework/build.sbt (hunk 2 above)
Comment (s-vitaliy) [reply framework/arcane-framework/build.sbt:8]: done

File: framework/arcane-framework/.idea/modules/arcane-framework.arcane-framework-build.iml (hunk 1 above)
Comment (s-vitaliy) [reply framework/arcane-framework/.idea/modules/arcane-framework.arcane-framework-build.iml]: done

**Pull request merged**